mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -260,126 +261,123 @@ struct MILARewritePattern : public OpRewritePattern<linalg::GenericOp> {
 };
 } // end anonymous namespace
 
-/// If `inpMap` is a map of the form
-/// (d0, d1, ..., dk) -> (d(i0), d(i1), ..., d(ik))
-/// where thi i's don't make it the identity map, wrap `inp` in a
-/// rock.transform that corresponds to the map, and returns the indexing map
-/// that is the result of applying the permutation. If no permutation is needed,
-/// returns its inputs.
-static std::tuple<Value, AffineMap>
-makeTransposeTransform(PatternRewriter &b, Value inp, AffineMap inpMap) {
-  if (!inpMap.isMinorIdentityWithBroadcasting()) {
-    // permumation[i] says where the map output i should be sent.
-    SmallVector<uint32_t> permutation;
-    if (inpMap.isPermutationOfMinorIdentityWithBroadcasting(permutation)) {
-      Location loc = inp.getLoc();
-      MemRefType inputType = inp.getType().cast<MemRefType>();
-      ArrayRef<int64_t> inputShape = inputType.getShape();
-      LLVM_DEBUG(llvm::dbgs()
-                 << "Transpose input type : " << inputType << "\n");
-      BottomUpTMBuilder permuteBuilder(b, inputShape, loc);
-
-      SmallVector<uint32_t> identityIdxs;
-      identityIdxs.reserve(inputShape.size());
-      for (uint32_t idx = 0, e = inputShape.size(); idx < e; ++idx)
-        identityIdxs.push_back(idx);
-
-      permuteBuilder.passThrough(permutation, identityIdxs);
-      TransformMapAttr permuteAttr = permuteBuilder.get();
-      Value ret = b.create<TransformOp>(loc, inp, permuteAttr);
-      AffineMap composed = permuteAttr.getMap().getAffineMap().compose(inpMap);
-      LLVM_DEBUG(llvm::dbgs() << "indexing = " << inpMap << " then transform "
-                              << permuteAttr.getMap().getAffineMap() << " is "
-                              << composed << "\n");
-      return {ret, composed};
+// This function will create a permutation that will permute the originalMap to
+// be a MinorIdentityWithBroadcast. This is used to add a permutation later in
+// the chain.
+// e.g. :
+// (d0, d1, d2, d4) -> (0, d1) was the map
+// This function will return a permutation : [0, 3, 2, 1] s.t.
+// apply it to the original map would result in
+// (d0, d4, d2, d1) -> (0, d1) in effect.
```

Comment:
Ok, hold on, I'm slightly confused.

The permutation map is, as I understand it `(d0, d1, d2, d4) -> (d0, d4, d2, d1)`
Then if, after that, I compose in the broadcast, I get the final map `(0, d4)`?

... right, which is a minor identity with broadcast. 

And the reason we need this is because we're actually going to apply the _inverse_ of that permutation to the GEMM coordinates in order to get correct indexing for the extra argument. 

That is, a function later, with the transpose/broadcast handling, we can describe the three sections as follows

/// Broadcast those dimensions that the original linalg.generic map specifies are broadcast and collect their locations, accounting for the leading dimensions not represented in that map but which are present in the gemm coordinates

/// Then, add dimensions that are present in the writeback coordinates but are not present in the additional fusion argument with matching sizes. This, combined with the previous step, ensures that the view of the fusion argument has the same dimensions as the gemm output, though they are not necessarily in the same order.

/// Permute the dimensions of the fusion argument to match those of the gemm writeback by applying the inverse of the permutation that would have made the original indexing map into a minor identity with broadcast. The inverse of that permutation takes the gemm writeback coordinates and scatters them into positions that match the non-identity indexing pattern of the fusion argument.

Reply:
Yep! well explained.

So I just copied them 😄